SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUpstreamChecker.java

```diff
@@ -0,0 +1,80 @@
+import com.hubspot.baragon.models.UpstreamInfo;
+import com.hubspot.mesos.protos.MesosParameter;
+import com.hubspot.singularity.SingularityTask;
+import com.hubspot.singularity.data.TaskManager;
+import com.hubspot.singularity.helpers.MesosUtils;
+import com.hubspot.singularity.hooks.LoadBalancerClient;
+import com.hubspot.singularity.hooks.LoadBalancerClientImpl;
+
+@Singleton
+public class SingularityUpstreamChecker {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LoadBalancerClient.class);
+
+  private final LoadBalancerClientImpl lbClient;
+  private final TaskManager taskManager;
+
+  @Inject
+  public SingularityUpstreamChecker(LoadBalancerClientImpl lbClient, TaskManager taskManager) {
+    this.lbClient = lbClient;
+    this.taskManager = taskManager;
+  }
+
+  private UpstreamInfo taskToUpstreamInfo (SingularityTask task, Optional<String> loadBalancerUpstreamGroup) {
+    UpstreamInfo upstreamInfo = null; //TODO: what is a better way to do this? We cannot initialize an empty upstream
+    final Optional<Long> maybeLoadBalancerPort = MesosUtils.getPortByIndex(lbClient.getMesosProtosUtils().toResourceList(task.getMesosTask().getResources()), task.getTaskRequest().getDeploy().getLoadBalancerPortIndex().or(0));
+    if (maybeLoadBalancerPort.isPresent()) {
+      final String host = task.getHostname();
+      final long port = maybeLoadBalancerPort.get();
+      String upstream = String.format("%s:%d", host, port);
+      final Optional<String> requestId = Optional.of(task.getTaskRequest().getRequest().getId());
+      Optional<String> group = loadBalancerUpstreamGroup;
+      if (lbClient.getTaskLabelForLoadBalancerUpstreamGroup().isPresent()) {
+        for (MesosParameter label : task.getMesosTask().getLabels().getLabels()) {
+          if (label.hasKey() && label.getKey().equals(lbClient.getTaskLabelForLoadBalancerUpstreamGroup().get()) && label.hasValue()) {
+            group = Optional.of(label.getValue());
+            break;
+          }
+        }
+      }
+      upstreamInfo = new UpstreamInfo(upstream, requestId, task.getRackId(), Optional.<String>absent(), group);
+    } else {
+      LOG.warn("Task {} is missing port, not passed to upstreams({})", task.getTaskId(), task);
+    }
+    return upstreamInfo;
+  }
+
+  public Collection<UpstreamInfo> getUpstreamsFromActiveTasks() {
+    final List<UpstreamInfo> upstreams = Lists.newArrayListWithCapacity(taskManager.getNumActiveTasks());
+    for (SingularityTask task: taskManager.getActiveTasks()){
+      final Optional<String> loadBalancerUpstreamGroup = task.getTaskRequest().getDeploy().getLoadBalancerUpstreamGroup();
+      final UpstreamInfo upstream = taskToUpstreamInfo(task, loadBalancerUpstreamGroup);
+      if (upstream.equals(null)){
+        upstreams.add(upstream);
+      }
+    }
+    return upstreams;
+  }
+
+  private void syncUpstreams( ) {
+    Collection<UpstreamInfo> upstreamsInBaragon = lbClient.getUpstreams();
```

Comment:
While it's more api calls, we should probably do this one service at a time for a few reasons:
- We want to compare upstreams within the context of each service. Right now the PR is compiling a list of all upstreams across all tasks/requests. It's very possible that host1:port1 is valid for SomethingWeb but not for SomethingElseWeb (in which case we would remove it from the one it is no longer valid for). We need that context of which request the upstream belongs to to make that distinction (the `request` field of the Baragon UpstreamInfo points to the BaragonRequest id, it is not the same as the SingularityRequest id)
- Updates are happening frequently from other Singularity pollers and we want to be able to lock the request and make sure we are the only ones currently modifying the state of the system. See examples of `runWithRequestLock` for how to lock by request id on the leader. In this way, we can ensure that the BaragonState we are operating on is not also being modified by another Singularity poller at the same time.

Reply:
Additional piece, we can limit the number of requests we loop over by first filtering to only those that have load balancing enabled (boolean property on SingularityRequest object)